### PR TITLE
Reduce runtime of Go Encode() by another 25%

### PR DESCRIPTION
**Pull request by willbeason**
This is (another) performance optimization change for the Go implementation of Encode(), which reduces the runtime by about 25%. This is the additional performance improvement I mentioned in #566.

There are two main changes here to accomplish this, either of which may be controversial as they do harm the code's readability. I have attempted to mitigate this with refactorings, and verified the extracted functions are inlined to avoid the additional overhead of function calls.

The first change is that all loops are unrolled. Given that all loops are unconditionally executed a constant number of times, this can be done without introducing any additional branches to the code.

The second change is that the precision rounding logic is modified to require zero divisions. This new method is significantly faster, but may change the final digits of codes requiring sub-centimeter precision (I don't believe this library has this as a supported use case?). There are no cases in the test suite in which this causes a difference, but if desired I can likely find such an edge case.

Before:

```
will@Janeway:~/GolandProjects/open-location-code/go$ go test --run=NONE --bench=Encode --benchtime=10s -cpuprofile=/home/will/cpu2.out
goos: linux
goarch: amd64
pkg: github.com/google/open-location-code/go
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor 
BenchmarkEncode-64    	186547320	        64.70 ns/op	      16 B/op	       1 allocs/op
PASS
ok  	github.com/google/open-location-code/go	22.222s
```

After:

```
will@Janeway:~/GolandProjects/open-location-code/go$ git checkout -
Switched to branch 'optimize-2'
will@Janeway:~/GolandProjects/open-location-code/go$ go test --run=NONE --bench=Encode --benchtime=10s -cpuprofile=/home/will/cpu.out
goos: linux
goarch: amd64
pkg: github.com/google/open-location-code/go
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor 
BenchmarkEncode-64    	233725107	        49.01 ns/op	      16 B/op	       1 allocs/op
PASS
ok  	github.com/google/open-location-code/go	21.471s
```

CPU profile of before:
profile002

CPU profile of after:
profile001

### CODE REVIEW

File: go/encode.go [hunk 1]
```diff
@@ -122,6 +105,52 @@ func Encode(lat, lng float64, codeLen int) string {
 	return string(code[:codeLen]) + strings.Repeat(string(Padding), sepPos-codeLen) + string(Separator)
 }
 
+func roundLatLngToInts(lat, lng float64) (int64, int64) {
```
[the remark this answers is not in the document]
Comment (willbeason) [reply go/encode.go:108]: Sounds good, will do. Responding to these as I take action - I'll resolve and send a PR message once I'm done.

File: go/encode.go [hunk 2]
```diff
@@ -63,56 +55,47 @@ func Encode(lat, lng float64, codeLen int) string {
 	// Use a char array so we can build it up from the end digits, without having
 	// to keep reallocating strings.
 	var code [16]byte
```
[the remark this answers is not in the document]
Comment (willbeason) [reply go/encode.go:57]: Nice!

File: go/encode.go [hunk 3]
```diff
@@ -63,56 +55,47 @@ func Encode(lat, lng float64, codeLen int) string {
...
 	} else {
 		latVal /= gridLatFullValue
 		lngVal /= gridLngFullValue
 	}
 
-	// Compute the pair after the Separator as a special case rather than
-	// introduce an if statement to the loop which will only be executed once.
-	// This also allows us to remove two unnecessary divides at the end of the loop.
+	// The first pair after the separator is a special case as it handles the
+	// transition from the grid.
```
[the remark this answers is not in the document]
Comment (willbeason) [reply go/encode.go:78]: Sounds good.

File: go/encode.go [hunk 4]
```diff
@@ -122,6 +105,52 @@ func Encode(lat, lng float64, codeLen int) string {
...
+}
+
+func latLngGrid(latVal, lngVal int64) (byte, int64, int64) {
+	latDigit := latVal % int64(gridRows)
+	lngDigit := lngVal % int64(gridCols)
+	ndx := latDigit*gridCols + lngDigit
+	return Alphabet[ndx], latVal / int64(gridRows), lngVal / int64(gridCols)
+}
+
+func latIter(latVal int64) (byte, int64) {
```
[the remark this answers is not in the document]
Comment (willbeason) [reply go/encode.go:142]: Good point on "Iter" - I originally wrote this before Iters were released.

**Comment by willbeason** [comment]
Okay, improved comments and renamed functions as requested.
